staging/src/k8s.io/apiserver/pkg/server/graceful_shutdown.go

```diff
@@ -0,0 +1,307 @@
+    - the HTTP Server waits gracefully for existing requests to complete
+      up to '10s' (it's hard coded right now)
+	- any new request arriving on a new TCP socket is denied with
+      a network error similar to 'connection refused'
+	- active long running requests will receive a GOAWAY here.
+
+*/
+
+// ShutdownEvents provides an abstraction of the events that transpire
+// during a shutdown. This abstraction makes it easy for us to write
+// unit tests that can verify expected graceful termination behavior
+//
+// GenericAPIServer can wait for these events to fire and
+// take action appropriately.
+type ShutdownEvents interface {
+	// ShutdownInitiated event fires when an apiserver shutdown has been initiated.
+	// It fires when the `stopCh` provided by the main goroutine
+	// receives a KILL signal and is closed as a consequence.
+	ShutdownInitiated() <-chan struct{}
+
+	// AfterShutdownDelayDuration event fires as soon as ShutdownDelayDuration has elapsed
+	// since the ShutdownInitiated event.
+	//
+	// ShutdownDelayDuration allows the apiserver to delay shutdown for some time.
+	AfterShutdownDelayDuration() <-chan struct{}
+
+	// NotAcceptingNewRequest event fires when the apiserver should not accept any
+	// new incoming request(s).
+	//
+	// Any incoming request after this event fires should be replied with with a 429
+	// and the response headers: 'Retry-After: 1` and 'Connection: close'
+	NotAcceptingNewRequest() <-chan struct{}
+
+	// InitiateHTTPServerShutdown event fires when it is time for the
+	// apiserver to initiate shutdown of the HTTP Server.
+	//
+	// When this event fires depends on the value of 'DelayServerShutdown'.
+	//  - true: this event will fire when all existing request(s)
+	//    in flight have drained.
+	//  - false: this event fires as soon as ShutdownDelayDuration elapses.
+	InitiateHTTPServerShutdown() <-chan struct{}
+
+	// InFlightRequestsDrained event fires when the existing requests
+	// in flight have completed.
+	// This can be used as signal to shut down the audit backends
+	InFlightRequestsDrained() <-chan struct{}
+}
+
+// GracefulShutdown is used by the GenericAPIServer to coordinate
+// graceful termination of the apiserver instance.
+type GracefulShutdown interface {
+	// Start accepts the 'StopCh' provided by the primary goroutine
+	// and uses it to fire the shutdown events at appropriate time.
+	// It returns ShutdownEvents that the caller can use to get notified
+	// when the shutdown events take place
+	//
+	// The GenericAPIServer should call Start once and then use the
+	// ShutdownEvents returned to coordinate a graceful termination.
+	//
+	// Start should be an idempotent call, calling it multiple times
+	// should not have any adverse effects.
+	Start(stopCh <-chan struct{}) ShutdownEvents
+
+	// Events returns the ShutdownEvents associated. The caller can
+	// can use to get notified when the shutdown events take place.
+	Events() ShutdownEvents
```

Comment:
do we really need Events() and the return value of Start for the same thing? Looks like these two overlap.